### PR TITLE
[WIP] Refactored argument resolving out of the ControllerResolver

### PR DESCRIPTION
**Pull request by wouterj**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | yes |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | #1547, #10710 |
| License | MIT |
| Doc PR | tbd |
# Todo
- [ ] Write docs
- [ ] Make BC
- [ ] Add priority
- [ ] Review
# BC

I've done my best to make this BC. It is now fully BC except from one case: If the logic in `ControllerResolver#getArguments()` is no longer used. If one would have used that method for custom argument resolving logic, it is no ignored. There should be a way that we can support both methods at the same time (in other words, create an argument resolver which uses the controller resolver).
# Thanks

A huge thanks to @Iltar for giving me the idea and helping me setting up this PR.

### CODE REVIEW

**Comment by WouterJ** [comment]
I just revisited this PR for 3.0, as it was impossible to make a BC pull request. I've fixed the comments.

This could be made so generic that it will work for any callable (not just controllers). This would cover #10710

**Comment by WouterJ** [comment]
What do core devs think about this PR?

1. Continue and make it decoupled from controllers
1. Continue the current direction, just as a replacement of SensioFrameworkExtraBundle's feature
1. Stop worrying about this PR

**Comment by fabpot** [comment]
@WouterJ I would say 2

**Comment by WouterJ** [comment]
Why can't HHVM find the compiler pass class?

File: src/Symfony/Component/HttpKernel/Controller/ArgumentResolverManager.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ * resolve all controller arguments.
+ *
+ * @author Wouter J <wouter@wouterj.nl>
+ */
+class ArgumentResolverManager
+{
+    /**
+     * @var ArgumentResolverInterface[]
+     */
+    protected $resolvers = array();
```
Comment (stof) [inline src/Symfony/Component/HttpKernel/Controller/ArgumentResolverManager.php]: should be private

**Comment by stof** [comment]
@WouterJ I just thought about a BC way introduce the ArgumentResolver concept in Symfony: inject the ArgumentResolverManager inside the ControllerResolver for now (with a default value matching the current behavior) to be used in its existing ``getArguments`` method.

This would allow to start using this extension point already.

Building an ArgumentResolver which would call the ControllerResolver old API would not allow to preserve BC, because you would not have BC for ``parent::getArguments()``

**Comment by WouterJ** [comment]
@stof this isn't FC for standalone usage, is it? As you have to pass the argument resolver manager with custom resolvers to ControllerResolver in 2.8 and to AppKernel in 3.0.

**Comment by WouterJ** [comment]
Hmm, we can of course still pass the argument resolver manager to AppKernel and introduce a setArgumentResolverManager method in the ControllerResolver (which is called in AppKernel). This would fix that.

**Comment by WouterJ** [comment]
Closing in favor of https://github.com/symfony/symfony/pull/14971

**Pull request closed without merging**